File: go/eval.go
package amino

import "strings"

// Func implements a schema-declared function on the in-process target.
type Func func(args ...any) (any, error)

type evalFn func(rec map[string]any, fns map[string]Func) (any, error)

// compileNode turns the typed AST into a closure. Errors during evaluation
// make the rule false; see CompiledRules.
func compileNode(n Node) evalFn {
	switch node := n.(type) {
	case *Literal:
		v := node.Value
		return func(map[string]any, map[string]Func) (any, error) { return v, nil }

	case *Variable:
		parts := strings.Split(node.Name, ".")
		name := node.Name
		return func(rec map[string]any, _ map[string]Func) (any, error) {
			var cur any = rec
			for _, part := range parts {
				m, ok := cur.(map[string]any)
				if !ok {
					return nil, fieldError(CodeEvaluation, name, "field %q not found", name)
				}
				v, ok := m[part]
				if !ok {
					return nil, fieldError(CodeEvaluation, name, "field %q not found", name)
				}
				cur = v
			}
			return cur, nil
		}

	case *Unary:
		operand := compileNode(node.Operand)
		if node.Op == "not" {
			return func(rec map[string]any, fns map[string]Func) (any, error) {
				v, err := operand(rec, fns)
				if err != nil {
					return nil, err
				}
				return !truthy(v), nil
			}
		}
		fn := node.Def.Fn
		return func(rec map[string]any, fns map[string]Func) (any, error) {
			v, err := operand(rec, fns)
			if err != nil {
				return nil, err
			}
			return fn(v)
		}

	case *Binary:
		left := compileNode(node.Left)
		right := compileNode(node.Right)
		switch node.Op {
		case "and":
			return func(rec map[string]any, fns map[string]Func) (any, error) {
				lv, err := left(rec, fns)
				if err != nil {
					return false, nil
				}
				if !truthy(lv) {
					return false, nil
				}
				rv, err := right(rec, fns)
				if err != nil {
					return nil, err
				}
				return truthy(rv), nil
			}
		case "or":
			return func(rec map[string]any, fns map[string]Func) (any, error) {
				lv, err := left(rec, fns)
				if err == nil && truthy(lv) {
					return true, nil
				}
				rv, err := right(rec, fns)
				if err != nil {
					return false, nil
				}
				return truthy(rv), nil
			}
		}
		fn := node.Def.Fn
		return func(rec map[string]any, fns map[string]Func) (any, error) {
			lv, err := left(rec, fns)
			if err != nil {
				return nil, err
			}
			rv, err := right(rec, fns)
			if err != nil {
				return nil, err
			}
			return fn(lv, rv)
		}

	case *Call:
		args := make([]evalFn, len(node.Args))
		for i, a := range node.Args {
			args[i] = compileNode(a)
		}
		name := node.Name
		return func(rec map[string]any, fns map[string]Func) (any, error) {
			fn, ok := fns[name]
			if !ok {
				return nil, newError(CodeEvaluation, "function %q has no implementation", name)
			}
			vals := make([]any, len(args))
			for i, a := range args {
				v, err := a(rec, fns)
				if err != nil {
					return nil, err
				}
				vals[i] = v
			}
			return fn(vals...)
		}
	}
	panic("unreachable node type")
}

File: typescript/src/eval.ts
import type { Node } from "./ast.js";
import { AminoError } from "./errors.js";
import { truthy } from "./values.js";

/** Implements a schema-declared function on the in-process target. */
export type Func = (...args: unknown[]) => unknown;

export type EvalFn = (record: Record<string, unknown>, fns: Record<string, Func>) => unknown;

/** Turns the typed AST into a closure. An error thrown during evaluation makes the rule false. */
export function compileNode(n: Node): EvalFn {
  switch (n.node) {
    case "literal": {
      const v = n.value;
      return () => v;
    }
    case "variable": {
      const parts = n.name.split(".");
      const name = n.name;
      return (record) => {
        let cur: unknown = record;
        for (const part of parts) {
          if (cur === null || typeof cur !== "object" || Array.isArray(cur) || !(part in (cur as object))) {
            throw new AminoError("evaluation", `field "${name}" not found`, name);
          }
          cur = (cur as Record<string, unknown>)[part];
        }
        return cur;
      };
    }
    case "unary": {
      const operand = compileNode(n.operand);
      if (n.op === "not") return (r, f) => !truthy(operand(r, f));
      const fn = n.def.fn;
      if (!fn) throw new AminoError("config", `operator "${n.op}" has no implementation`);
      return (r, f) => fn(operand(r, f));
    }
    case "binary": {
      const left = compileNode(n.left);
      const right = compileNode(n.right);
      if (n.op === "and") {
        return (r, f) => {
          let lv: unknown;
          try {
            lv = left(r, f);
          } catch {
            return false;
          }
          if (!truthy(lv)) return false;
          return truthy(right(r, f));
        };
      }
      if (n.op === "or") {
        return (r, f) => {
          try {
            if (truthy(left(r, f))) return true;
          } catch {
            // a failing left side is treated as false
          }
          try {
            return truthy(right(r, f));
          } catch {
            return false;
          }
        };
      }
      const fn = n.def.fn;
      if (!fn) throw new AminoError("config", `operator "${n.op}" has no implementation`);
      return (r, f) => fn(left(r, f), right(r, f));
    }
    case "call": {
      const args = n.args.map(compileNode);
      const name = n.name;
      return (r, f) => {
        const fn = f[name];
        if (!fn) throw new AminoError("evaluation", `function "${name}" has no implementation`);
        return fn(...args.map((a) => a(r, f)));
      };
    }
  }
}
